tap-version-14-specification.md

```diff
@@ -0,0 +1,1161 @@
+
+For example, a test file's output might look like:
+
+```tap
+TAP version 14
+1..4
+ok 1 - Input file opened
+not ok 2 - First line of the input valid
+  ---
+  message: 'First line invalid'
+  severity: fail
+  data:
+    got: 'Flirble'
+    expect: 'Fnible'
+  ...
+ok 3 - Read the rest of the file
+not ok 4 - Summarized correctly # TODO Not written yet
+  ---
+  message: "Can't make summary yet"
+  severity: todo
+  ...
+```
+
+## Harness Behavior
+
+In this document, the "harness" is any program analyzing TAP output.
+
+Typically this will be a test framework's runner program, but may also be a
+programmatic parser, parent test object, or test result reporter.
+
+In a typical TAP implementation, tests are programs that ouput TAP data
+according to this specification.  The "test harness" reads TAP output from
+these programs, and handles it in some way.
+
+A harness that is collecting output from a test program _should_ read and
+interpret TAP from the process's standard output, not standard error.
+(Handling of test standard error is implementation-specific.)
+
+A Harness _should_ normalize line endings by replacing any instances of
+`\r\n` or `\r` in the TAP document with `\n`.
+
+A harness _should_ treat a test program as a failed test if:
+
+* The TAP output lines indicate test failure, or
+* The TAP output of the process is invalid in a way that is not
+  recoverable, or
+* The exit code of the test program is not 0 (including test programs
+  killed by a fatal Unix signal).
+
+If one or more test programs are considered failures, then a TAP Harness
+_should_ indicate failure to the user in whatever means are appropriate.
+For example, a command-line test runner might exit with a non-zero status
+code, while a web-based continuous integration system might put a red
+"FAILED" notice in the html output.
+
+Note that some of the above guidance may not apply if the Harness is
+interpreting TAP from a different sort of text stream, or for a purpose
+other than actually running tests.  For example, a Harness may be used to
+analyze the recorded output of past test runs and provide data about them.
+
+## Document Structure
+
+TAP14 producers _must_ encode TAP data using the UTF-8 encoding.
+
+Harnesses _should_ interpret all TAP streams using UTF-8.  Harnesses _may_
+provide a mechanism for using other encodings explicitly, if needed for
+backwards compatibility.
+
+### Version
+
+To indicate that this is TAP14 the first line _must_ be
+
+```tap
+TAP version 14
+```
+
+Harnesses _may_ interpret ostensibly
+[TAP13](https://testanything.org/tap-version-13-specification.html) streams
+as TAP14, as this specification is compatible with observed behavior of
+existing TAP13 consumers and producers.  That is, they _may_ treat this as
+a valid Version line while parsing TAP14:
+
+```tap
+TAP version 13
+```
+
+Harnesses _may_ treat any TAP stream lacking a version as a failed test.
```

Comment:
I implemented this in Meson as a passing test which verbosely warns you not to do this. I got a response:

https://github.com/mesonbuild/meson/pull/11186#issuecomment-1362809587

> Meson is able to parse TAP12 and TAP13, therefore it does not have to warn if the version line is missing.

> IMO the "Harnesses may treat any TAP stream lacking a version as a failed test" is just silly. The point of TAP is exactly that it is easy to produce without relying on a library that handles the versioning and formatting for you, so good luck enforcing that with dozens of TAP producers.

Thoughts?

Reply:
More importantly:

> In fact the test-anything.org page says:
> 
> > Here’s what a TAP test stream looks like:
> > ```
> > 1..4
> > ok 1 - Input file opened
> > not ok 2 - First line of the input valid
> > ok 3 - Read the rest of the file
> > not ok 4 - Summarized correctly # TODO Not written yet
> > ```

Which regardless of anything else seems to be a problem -- the homepage of https://testanything.org/ describes a TAP stream without the soft-mandatory version, so it is valid TAP 12 but a TAP 14 harness can report it as a test failure. Maybe this should include a version...

Reply:
Thanks for bringing it up here @eli-schwartz. In my opinion the correct reading of the spec is:

- a TAP harness MUST treat a missing version line as version 12

- a TAP harness MAY reject versions below any limit they would like to apply (even though it's not a good idea :))

The latter does imply rejecting input without a version line (which puts the limit at version 13), but does not mean that harnesses should warn if the version line is absent.

Reply:
Let's discuss this in a new issue, I think it's a useful subject to dig into and get more opinions on.
https://github.com/TestAnything/Specification/issues/41